### ISSUE TITLE
When I use your checkpoint "model_brats.pt" to generate images, I can only obtain low-quality MRI sample

### ISSUE THREAD

**Issue by TUAT-Novice** [mobaidoctor/med-ddpm#28]
Thank you so much for your outstanding work for the repository. 

I want to use your brats model weights to generate some MRI data, but find that the artificial samples are with low quality. In my program, I: 
1. git clone your github repository, then prepare the required environment for the codes;
2. download the model_brats.pt checkpoint from the Google Drive (put in ./model/);
3. download the brats2021 dataset, then run the "preprocess_brats_data.py" script to prepare and preload the dataset (put in ./dataset/brats2021/);
4. modify the "inputfolder" parameter, then run the "sample_brats.py" script to generate the artificial MRI with the brats2021 segmentation mask saving in directory ./dataset/brats2021/seg/

I am not very clear about what went wrong in the process. If possible, I hope you can provide some advice on this matter. Thank you so much for your kindest help.

**Comment by mobaidoctor** [comment]
@TUAT-Novice Thank you for your interest in our work and your inquiry. We apologize for the delayed response.
If you use our pretrained weights and the provided preprocessing script for the segmentation mask during inference, it should work without issues. Could you please share your results with us? This will help us investigate the problem in detail. Thank you!

**Comment by supgy** [comment]
> > @TUAT-Novice Thank you for your interest in our work and your inquiry. We apologize for the delayed response. If you use our pretrained weights and the provided preprocessing script for the segmentation mask during inference, it should work without issues. Could you please share your results with us? This will help us investigate the problem in detail. Thank you!
> 
> Of course! I have sent the results to you through email. Thank you so much for your help!

Hello， I tried the same steps as you, but got the result like follows, do you know why did this result occur?
![image](https://github.com/user-attachments/assets/bd48f84e-f592-498e-9abd-dd1537dcd12d)

**Comment by TUAT-Novice** [comment]
> > > @TUAT-Novice Thank you for your interest in our work and your inquiry. We apologize for the delayed response. If you use our pretrained weights and the provided preprocessing script for the segmentation mask during inference, it should work without issues. Could you please share your results with us? This will help us investigate the problem in detail. Thank you!
> > 
> > 
> > Of course! I have sent the results to you through email. Thank you so much for your help!
> 
> Hello， I tried the same steps as you, but got the result like follows, do you know why did this result occur? ![image](https://private-user-images.githubusercontent.com/144404526/348967097-bd48f84e-f592-498e-9abd-dd1537dcd12d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjExMTYyMjUsIm5iZiI6MTcyMTExNTkyNSwicGF0aCI6Ii8xNDQ0MDQ1MjYvMzQ4OTY3MDk3LWJkNDhmODRlLWY1OTItNDk4ZS05YWJkLWRkMTUzN2RjZDEyZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxNlQwNzQ1MjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zMGE5NGRhYWM0YzQxN2Q3MWQ0YzVmMzllMzE5YTcwYWNmMmQyNTY2NDk3MjNkZjg0NjM1MWM0ZDdkZTg3Y2Y1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.bc_5rYB0JHc6a3If0Au6WwKjtg3XPjwHVqkiXtpGTqc)

I got the same results as you. But I can not solve the problem so far.

**Comment by supgy** [comment]
> > > > @TUAT-Novice Thank you for your interest in our work and your inquiry. We apologize for the delayed response. If you use our pretrained weights and the provided preprocessing script for the segmentation mask during inference, it should work without issues. Could you please share your results with us? This will help us investigate the problem in detail. Thank you!
> > > 
> > > 
> > > Of course! I have sent the results to you through email. Thank you so much for your help!
> > 
> > 
> > Hello， I tried the same steps as you, but got the result like follows, do you know why did this result occur? ![image](https://private-user-images.githubusercontent.com/144404526/348967097-bd48f84e-f592-498e-9abd-dd1537dcd12d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjExMTYyMjUsIm5iZiI6MTcyMTExNTkyNSwicGF0aCI6Ii8xNDQ0MDQ1MjYvMzQ4OTY3MDk3LWJkNDhmODRlLWY1OTItNDk4ZS05YWJkLWRkMTUzN2RjZDEyZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxNlQwNzQ1MjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zMGE5NGRhYWM0YzQxN2Q3MWQ0YzVmMzllMzE5YTcwYWNmMmQyNTY2NDk3MjNkZjg0NjM1MWM0ZDdkZTg3Y2Y1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.bc_5rYB0JHc6a3If0Au6WwKjtg3XPjwHVqkiXtpGTqc)
> 
> I got the same results as you. But I can not solve the problem so far.

All rightT_T, but using the author's original data can yield accurate results

**Comment by TUAT-Novice** [comment]
> > > > > @TUAT-Novice Thank you for your interest in our work and your inquiry. We apologize for the delayed response. If you use our pretrained weights and the provided preprocessing script for the segmentation mask during inference, it should work without issues. Could you please share your results with us? This will help us investigate the problem in detail. Thank you!
> > > > 
> > > > 
> > > > Of course! I have sent the results to you through email. Thank you so much for your help!
> > > 
> > > 
> > > Hello， I tried the same steps as you, but got the result like follows, do you know why did this result occur? ![image](https://private-user-images.githubusercontent.com/144404526/348967097-bd48f84e-f592-498e-9abd-dd1537dcd12d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjExMTYyMjUsIm5iZiI6MTcyMTExNTkyNSwicGF0aCI6Ii8xNDQ0MDQ1MjYvMzQ4OTY3MDk3LWJkNDhmODRlLWY1OTItNDk4ZS05YWJkLWRkMTUzN2RjZDEyZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxNlQwNzQ1MjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zMGE5NGRhYWM0YzQxN2Q3MWQ0YzVmMzllMzE5YTcwYWNmMmQyNTY2NDk3MjNkZjg0NjM1MWM0ZDdkZTg3Y2Y1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.bc_5rYB0JHc6a3If0Au6WwKjtg3XPjwHVqkiXtpGTqc)
> > 
> > 
> > I got the same results as you. But I can not solve the problem so far.
> 
> All rightT_T, but using the author's original data can yield accurate results

Yes, I checked their .npy files and found their results are pretty good.

**Comment by mobaidoctor** [comment]
Hi, @TUAT-Novice and @supgy, thank you for your interest in our work. Our pre-trained BRATS model generates accurate images as it has been trained exclusively on carefully selected, high-quality BRATS images. Could you please send us the exact input mask image you used with the inference script?

@TUAT-Novice Sorry for the late response, we have been fully occupied with project tasks. We received your email, but it only contained the exported images. We need to verify the input mask you provided to the model. Please ensure that the mask image has been processed using the preprocess_seg function from the preprocess_brats_data.py file in our repository. Thank you.

**Comment by TUAT-Novice** [comment]
> Hi, @TUAT-Novice and @supgy, thank you for your interest in our work. Our pre-trained BRATS model generates accurate images as it has been trained exclusively on carefully selected, high-quality BRATS images. Could you please send us the exact input mask image you used with the inference script?
> 
> @TUAT-Novice Sorry for the late response, we have been fully occupied with project tasks. We received your email, but it only contained the exported images. We need to verify the input mask you provided to the model. Please ensure that the mask image has been processed using the preprocess_seg function from the preprocess_brats_data.py file in our repository. Thank you.

Thank you so much for your reply. I just confirmed the email, and I found the output segmentation under the "./exports/seg/" directory. I'm not sure if such output is the same as the input mask. If they are not the same, I will save the preprocessed mask as an "npy" file and send it to you via email again. Thank you very much for your response.

**Comment by TUAT-Novice** [comment]
> Hi, @TUAT-Novice and @supgy, thank you for your interest in our work. Our pre-trained BRATS model generates accurate images as it has been trained exclusively on carefully selected, high-quality BRATS images. Could you please send us the exact input mask image you used with the inference script?
> 
> @TUAT-Novice Sorry for the late response, we have been fully occupied with project tasks. We received your email, but it only contained the exported images. We need to verify the input mask you provided to the model. Please ensure that the mask image has been processed using the preprocess_seg function from the preprocess_brats_data.py file in our repository. Thank you.

For the mask image, I think I have preprocessed with the preprocess_seg function, since I preprocess the whole brats dataset with the main function you provide in the "preprocess_brats_data.py" script, with the source data file and the target file as input. Then, I use the target file (e.g. "../dataset/brats2021/seg") as a novel "inputfolder" for sampling.

**Comment by mobaidoctor** [comment]
@TUAT-Novice If you used our preprocessing function, your input mask image must be in .nii.gz format because our model accepts only NIfTI images, not .npy files. The sample_brats code contains a processMsk function that modifies the exported images back to the original BRATS dataset format, such as the image size of (240, 240, 155), and adjusts the segmentation mask to the original BRATS format without the brain shape. Therefore, I need to know what your input NIfTI file was for the sampling script. Please send me your input file, not the output. Thank you.

**Comment by TUAT-Novice** [comment]
> @TUAT-Novice If you used our preprocessing function, your input mask image must be in .nii.gz format because our model accepts only NIfTI images, not .npy files. The sample_brats code contains a processMsk function that modifies the exported images back to the original BRATS dataset format, such as the image size of (240, 240, 155), and adjusts the segmentation mask to the original BRATS format without the brain shape. Therefore, I need to know what your input NIfTI file was for the sampling script. Please send me your input file, not the output. Thank you.

I have sent you another email. Thank you very much!!!

**Comment by mobaidoctor** [comment]
@TUAT-Novice @supgy Thank you so much for letting us know about the problem. The strange output happened because your input mask did not include the brain shape label, which we added as a new label for the segmentation mask before using it in the model. Then, we checked our preprocessing script and found a bug. When we combined our preprocessing functions into one file, we forgot to update the file path before the final preprocessing of segmentation masks. We are very sorry for the confusion this caused. We have fixed the bug and updated the script. You can now use it without any issues. We appreciate your feedback and support. Thank you again for bringing this to our attention.

![image](https://github.com/user-attachments/assets/6aa5e694-79b6-4228-8229-d903ec5276ad)

**Comment by TUAT-Novice** [comment]
> @TUAT-Novice @supgy Thank you so much for letting us know about the problem. The strange output happened because your input mask did not include the brain shape label, which we added as a new label for the segmentation mask before using it in the model. Then, we checked our preprocessing script and found a bug. When we combined our preprocessing functions into one file, we forgot to update the file path before the final preprocessing of segmentation masks. We are very sorry for the confusion this caused. We have fixed the bug and updated the script. You can now use it without any issues. We appreciate your feedback and support. Thank you again for bringing this to our attention.
> 
> ![image](https://private-user-images.githubusercontent.com/105937111/349088866-6aa5e694-79b6-4228-8229-d903ec5276ad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEyMTg0NzgsIm5iZiI6MTcyMTIxODE3OCwicGF0aCI6Ii8xMDU5MzcxMTEvMzQ5MDg4ODY2LTZhYTVlNjk0LTc5YjYtNDIyOC04MjI5LWQ5MDNlYzUyNzZhZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxN1QxMjA5MzhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yZTA1MmI0ZTJmODA3MzY3M2QwNjFjNDA0YjFhYTJhOTg5MjY0ZWRhOWE4MGM3OGFlYjZlYmY0MjZlMjkyOGQ5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ndg5HIzCgJK1tDzBVOPFrHuxnmKIzG6BWiziIrogs7c)

Thank you for your checking~~ I can receive the results correctly. I will close the ISSUE.
